sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UseUriInsteadOfString.cs

```diff
@@ -51,75 +53,143 @@ public sealed class UseUriInsteadOfString : SonarDiagnosticAnalyzer
+                SyntaxKind.InvocationExpression,
+                SyntaxKind.ObjectCreationExpression);
+        }
 
-            context.RegisterSyntaxNodeActionInNonGenerated(
-                c =>
+        private void VerifyDeclaration(SyntaxNodeAnalysisContext context)
+        {
+            var methodDeclaration = (BaseMethodDeclarationSyntax)context.Node;
+            var methodSymbol = context.SemanticModel.GetDeclaredSymbol(methodDeclaration);
+
+            if (methodDeclaration == null || methodSymbol == null || methodSymbol.IsOverride)
+            {
+                return;
+            }
+
+            VerifyReturnType(context, methodDeclaration, methodSymbol);
+
+            var stringUrlParams = GetStringUrlParamIndexes(methodSymbol);
+            if (!stringUrlParams.Any())
+            {
+                return;
+            }
+
+            var methodOverloads = FindOverloadsThatUseUriTypeInPlaceOfString(methodSymbol, stringUrlParams);
+            if (methodOverloads.Any())
+            {
+                var methodOverloadSet = new HashSet<IMethodSymbol>(methodOverloads);
+                if (!methodDeclaration.IsKind(SyntaxKind.ConstructorDeclaration) &&
+                    !SyntaxHelper.ContainsMethodInvocation(methodDeclaration, context.SemanticModel,
+                        mSyntax => true,
+                        mSymbol => methodOverloadSet.Contains(mSymbol)))
                 {
-                    var propertyDeclaration = (PropertyDeclarationSyntax)c.Node;
-                    var propertySymbol = c.SemanticModel.GetDeclaredSymbol(propertyDeclaration);
+                    context.ReportDiagnostic(Diagnostic.Create(rule_S3997, FindIdentifierLocation(methodDeclaration)));
+                }
+            }
+            else
+            {
+                foreach (int paramIdx in stringUrlParams)
+                {
+                    context.ReportDiagnostic(Diagnostic.Create(rule_S3994,
+                        methodDeclaration.ParameterList.Parameters[paramIdx].Type.GetLocation()));
+                }
+            }
+        }
 
-                    if (propertySymbol.Type.Is(KnownType.System_String) &&
-                        !propertySymbol.IsOverride &&
-                        NameContainsUri(propertySymbol.Name))
-                    {
-                        c.ReportDiagnostic(Diagnostic.Create(rule_S3996, propertyDeclaration.Type.GetLocation()));
-                    }
-                },
-                SyntaxKind.PropertyDeclaration);
+        private Location FindIdentifierLocation(BaseMethodDeclarationSyntax baseMethodDeclaration)
+        {
+            var location = (baseMethodDeclaration as MethodDeclarationSyntax)?.Identifier.GetLocation();
+            if (location == null)
+            {
+                location = (baseMethodDeclaration as ConstructorDeclarationSyntax)?.Identifier.GetLocation();
+            }
+
+            return location;
```

Comment:
This will probably be a bit better if written like:

        return (baseMethodDeclaration as MethodDeclarationSyntax)?.Identifier.GetLocation()
            ?? (baseMethodDeclaration as ConstructorDeclarationSyntax)?.Identifier.GetLocation();